skills_ml/algorithms/occupation_classifiers/train.py

```diff
@@ -0,0 +1,131 @@
+        self.grid_config = self.default_grid_config if grid_config is None else grid_config
+        self.cls_cv_result = {}
+        self.scores = scores
+        self.best_classifiers = {}
+        self.random_state_for_split = random_state_for_split
+
+    @property
+    def default_grid_config(self):
+        return {
+                'sklearn.ensemble.ExtraTreesClassifier': {
+                    'n_estimators': [100, 500],
+                    'criterion': ['gini', 'entropy'],
+                    'max_depth': [1, 5, 10, 20, 50],
+                    'max_features': ['sqrt', 'log2'],
+                    'min_samples_split': [2, 5, 10]
+                    },
+                'sklearn.ensemble.RandomForestClassifier': {
+                    'n_estimators': [100, 500],
+                    'criterion': ['gini', 'entropy'],
+                    'max_depth': [1, 5, 10, 20, 50],
+                    'max_features': ['sqrt', 'log2'],
+                    'min_samples_split': [2, 5, 10]
+                    }
+                }
+
+    def train(self):
+        """Fit a model to a training set. Works on any modeling class that
+        is vailable in this package's environment and implements .fit
+        """
+        X = self.matrix['X']
+        y = self.matrix['y']
+        for score in self.scores:
+            for class_path, parameter_config in self.grid_config.items():
+                module_name, class_name = class_path.rsplit(".", 1)
+                module = importlib.import_module(module_name)
+                cls = getattr(module, class_name)
+                kf = StratifiedKFold(n_splits=self.k_folds, random_state=self.random_state_for_split)
+
+                cls_cv = GridSearchCV(cls(), parameter_config, cv=kf, scoring=score, n_jobs=self.n_jobs)
+                cls_cv.fit(X, y)
+                self.cls_cv_result[score] = {class_name: cls_cv.cv_results_}
+                self.best_classifiers[score] = {class_name: cls_cv}
+
+    def save_result(self):
+        return NotImplementedError
+
+
+class SocEncoder(LabelEncoder):
+    def __init__(self, label_list):
+        self.fit(label_list)
+
+
+def create_training_set(job_postings_generator, embedding_model=None, target_variable="major_group",
+                        document_schema_fields=['description','experienceRequirements', 'qualifications', 'skills']):
+    """Create training set for occupation classifier from job postings generator and embedding model
+
+    Args:
+        job_postings_generator (generator): job posting collection
+        emebdding_model (skills_ml.algorithm.embedding.models): embedding model
+        document_schema_fields (list): fields to be included in the training data
+
+    Returns:
+        (dict) a dictionary of training data(X), label(y), the embedding_model, target_variable and soc_encoder
+    """
+    X = []
+    y = []
+
+    if target_variable == "major_group":
+        se = SocEncoder(list(majorgroupname.keys()))
```

Comment:
Not totally necessary to do any of this, but my thoughts on this section:

I think it would be cleaner to define these lists higher up, since it doesn't seem like it's just the trainer that uses them. At module level (like in occupation_classifiers.__init__.py)

Even better: Since the list and label transformer are closely coupled, put them in the same thing!

```
class TargetVariable(object):
  abstract name

class SOCMajorGroup(TargetVariable):
    name = 'major_group'

    def __init__(self):
        self.choices = list(majorgroupname.keys())
        self.encoder = SocEncoder(self.choices)
        self.transformer = lambda job_posting: self.encoder.transform([job_posting['onet_soc_code'][:2]])

```

With a little more thought you could even fit the filtering (if soc code doesn't start with 99, etc) into this framework.